### PR TITLE
Update pybind to 2.6.0

### PR DESCRIPTION
**Pull request by malfet**
Preserve PYBIND11 configuration options in `torch._C._PYBIND11_COMPILER_TYPE` and use them when building extensions

Also, use f-strings in `torch.utils.cpp_extension`

"Fixes" https://github.com/pytorch/pytorch/issues/46367

### CODE REVIEW

**Comment by henryiii** [comment]
So there are only two errors (same in two jobs, other jobs passed),

```
Traceback (most recent call last):
  File "test_cpp_extensions_jit.py", line 504, in test_cpp_frontend_module_python_inter_op
    verbose=True,
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 993, in load
    keep_intermediates=keep_intermediates)
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1209, in _jit_compile
    return _import_module_from_library(name, build_directory, is_python_module)
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/cpp_extension.py", line 1560, in _import_module_from_library
    return imp.load_module(module_name, file, path, description)
  File "/opt/conda/lib/python3.6/imp.py", line 243, in load_module
    return load_dynamic(name, filename, file)
  File "/opt/conda/lib/python3.6/imp.py", line 343, in load_dynamic
    return _load(spec)
  File "<frozen importlib._bootstrap>", line 684, in _load
  File "<frozen importlib._bootstrap>", line 658, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 571, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 922, in create_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
ImportError: generic_type: type "Net" referenced unknown base type "torch::nn::Module"
```

and

```
Traceback (most recent call last):
  File "test_cpp_extensions_jit.py", line 752, in test_warning
    warn_mod.foo(t, 1)
RuntimeError: Error with CPUDoubleType
```

Are those known or are they an issue of the upgrade?

**Comment by henryiii** [comment]
Interesting, the change causes macOS to fail similarly. That means the flags you are adding are being added when PyTorch is compiled, but not the JIT extension, I expect. Ideally, I think the "correct" solution would be to store the values of those two defines as added by pybind11 (which will be something like `"_clang"` or `"_gcc"` for `PYBIND11_COMPILER_TYPE`, and `"_libcpp"` or `"_libstdcpp"` for `PYBIND11_STDLIB`), and then set them to matching values via defines when compiling the JIT extension.

**Comment by henryiii** [comment]
Ah, I missed one. Can you also add `PYBIND11_BUILD_ABI` to an empty string too? That one was also added after the version currently there, just in a different commit. I think that will do it, and it's not my JIT idea.

**Comment by malfet** [comment]
I have the fix that addresses all the failures locally, will see if CI agrees with me.
But I'm still a bit uncertain, why those must be defined to empty string in the first place?

**Comment by henryiii** [comment]
@malfet you need to add `PYBIND11_BUILD_ABI` too. If you do, I think this will "pass" and prove that the only issue is the addition of these three items to the ABI string. I still think the correct fix would be to capture the values of `PYBIND11_COMPILER_TYPE`, `PYBIND11_STDLIB`, and `PYBIND11_BUILD_ABI` when PyTorch is compiled, and then query them and define them manually when running the JIT compile.

The old pybind11 used to build an ABI string like this:

```python
f"__pybind11_internals_v{PYBIND11_INTERNALS_VERSION}{PYBIND11_INTERNALS_KIND}{PYBIND11_BUILD_TYPE}__"
```

Since 2.4, it now builds one like this:
```python
f"__pybind11_internals_v{PYBIND11_INTERNALS_VERSION}{PYBIND11_INTERNALS_KIND}{PYBIND11_COMPILER_TYPE}{PYBIND11_STDLIB}{PYBIND11_BUILD_ABI}{PYBIND11_BUILD_TYPE}__"
```
This was mixing extensions built with different compilers sometimes segfaulted, so this was the "sledgehammer" solution. Ideally, compilers that are ABI compatible (at least for the parts that pybind11 uses in it's ABI) should have the same ABI string so they can interact, but for now, this was done to avoid hard to debug issues. If you know it works (such as is it was working before), then you can control the values of these new components.

I think in the failing test, `PYBIND11_COMPILER_TYPE ` is `"_clang"`, `PYBIND11_STDLIB` is `"_libcpp"`, and `PYBIND_BUILD_ABI` is `""`. When compiling the JIT, then `PYBIND11_COMPILER_TYPE ` is `"_gcc"`, `PYBIND11_STDLIB` is `"_libstdcxx"`, and `PYBIND_BUILD_ABI` is `f"_cxxabi{__GXX_ABI_VERSION}"`. If it can pass if we force all these to be the same, and the easiest way to do that is to set them to the empty string, then that will prove this is the issue.

The correct fix, I believe, will be to store these three values into the pytorch extension, say as "torch._detail.pybind11_compiler_type", etc, then to define the values to match when running the JIT compile step.

File: torch/CMakeLists.txt [hunk 1]
```diff
@@ -93,6 +93,9 @@ set(TORCH_PYTHON_COMPILE_OPTIONS)
 
 set(TORCH_PYTHON_LINK_FLAGS "")
 
+list(APPEND TORCH_PYTHON_COMPILE_DEFINITIONS PYBIND11_COMPILER_TYPE=\"\")
+list(APPEND TORCH_PYTHON_COMPILE_DEFINITIONS PYBIND11_STDLIB=\"\")
+
```
Comment (henryiii) [inline torch/CMakeLists.txt:98]:
```suggestion
list(APPEND TORCH_PYTHON_COMPILE_DEFINITIONS PYBIND11_BUILD_ABI =\"\")
```

File: torch/utils/cpp_extension.py [hunk 2]
```diff
@@ -364,6 +364,8 @@ def build_extensions(self) -> None:
                         extension.extra_compile_args[ext] = []
 
             self._add_compile_flag(extension, '-DTORCH_API_INCLUDE_EXTENSION_H')
+            self._add_compile_flag(extension, '-DPYBIND11_COMPILER_TYPE=""')
+            self._add_compile_flag(extension, '-DPYBIND11_STDLIB=""')
```
Comment (henryiii) [inline torch/utils/cpp_extension.py:368]:
```suggestion
            self._add_compile_flag(extension, '-DPYBIND11_STDLIB=""')
            self._add_compile_flag(extension, '-DPYBIND11_BUILD_ABI=""')
```

File: torch/utils/cpp_extension.py [hunk 3]
```diff
@@ -1594,6 +1596,8 @@ def _write_ninja_file_to_build_library(path,
 
     common_cflags = ['-DTORCH_EXTENSION_NAME={}'.format(name)]
     common_cflags.append('-DTORCH_API_INCLUDE_EXTENSION_H')
+    common_cflags.append('-DPYBIND11_COMPILER_TYPE=""')
+    common_cflags.append('-DPYBIND11_STDLIB=""')
```
Comment (henryiii) [inline torch/utils/cpp_extension.py:1600]:
```suggestion
    common_cflags.append('-DPYBIND11_STDLIB=""')
    common_cflags.append('-DPYBIND11_BUILD_ABI=""')
```

**Comment by ezyang** [comment]
Would have been nice to have the diff without the f-string refactor

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/46415#issuecomment-717967802 is a really good discussion and I'd expect some reference to it in the code

**Review by ezyang — APPROVED** [review]
This is a horrible hack but I understand the mechanism by which it works and agree that it is appropriate to restore old behavior.

I would like it to be hella more documented.

**Comment by malfet** [comment]
> Would have been nice to have the diff without the f-string refactor

Moved the refactors to https://github.com/pytorch/pytorch/pull/47025

**Comment by malfet** [comment]
Rebased PR on top of code refactors

**Comment by malfet** [comment]
Passing arguments with quotes via command line is non uniform on different platforms. On the other hand, I don't think there is much compiler variability on Windows

**Comment by henryiii** [comment]
> Passing arguments with quotes via command line is non uniform on different platforms

Yes, I've done it before here: https://github.com/pybind/python_example/blob/fc12654bf8915c3237ff4265e7e6889d228b5da7/src/main.cpp#L39 & https://github.com/pybind/python_example/blob/fc12654bf8915c3237ff4265e7e6889d228b5da7/setup.py#L24  but I agree, I don't think it's needed on Windows.

**Comment by malfet** [comment]
@henryiii Do you think it would make developers live easier if they are to be stringified on pybind11 side? (Can work on a PR if you want)

File: torch/utils/cpp_extension.py [hunk 4]
```diff
@@ -1590,6 +1595,24 @@ def _write_ninja_file_to_build_library(path,
...
+    # f"__pybind11_internals_v{PYBIND11_INTERNALS_VERSION}{PYBIND11_INTERNALS_KIND}{PYBIND11_COMPILER_TYPE}{PYBIND11_STDLIB}{PYBIND11_BUILD_ABI}{PYBIND11_BUILD_TYPE}__"
+    #
+    # This was done in order to further narrow down the chances of compiler ABI incompatibility
+    # that can cause a hard to debug segfaults.
+    # For PyTorch extensions we want to relax those restrictions and pass compiler, stdlib and abi properties
+    # captured during PyTorch native library compilation in torch/csrc/Module.cpp
+
+    for pname in ["COMPILER_TYPE", "STDLIB", "BUILD_ABI"]:
+        pval = getattr(torch._C, f"_PYBIND11_{pname}")
+        if pval is not None and not IS_WINDOWS:
```
Comment (peterjc123) [inline torch/utils/cpp_extension.py:1613]: Would you please tell me why we add `IS_WINDOWS` check here?

File: torch/utils/cpp_extension.py [hunk 5]
```diff
@@ -364,6 +364,11 @@ def build_extensions(self) -> None:
                         extension.extra_compile_args[ext] = []
 
             self._add_compile_flag(extension, '-DTORCH_API_INCLUDE_EXTENSION_H')
+            # See note [Pybind11 ABI constants]
+            for name in ["COMPILER_TYPE", "STDLIB", "BUILD_ABI"]:
+                val = getattr(torch._C, f"_PYBIND11_{name}")
+                if val is not None and not IS_WINDOWS:
```
Comment (peterjc123) [inline torch/utils/cpp_extension.py:370]: and here?

**Comment by peterjc123** [comment]
> Passing arguments with quotes via command line is non uniform on different platforms. On the other hand, I don't think there is much compiler variability on Windows

Well, then why don't we just don't quote all the args later on different platforms?

**Pull request closed without merging**